storage/bootstrap/bootstrapper/base.go

```diff
@@ -34,77 +35,118 @@ const (
+func (b *baseBootstrapper) Can(strategy bootstrap.Strategy) bool {
+	return b.src.Can(strategy)
+}
+
+func (b *baseBootstrapper) Bootstrap(shardsTimeRanges bootstrap.ShardTimeRanges) (bootstrap.Result, error) {
+	if shardsTimeRanges.IsEmpty() {
 		return nil, nil
 	}
 
-	availableRanges := bsb.s.GetAvailability(shard, targetRanges)
-	remainingRanges := targetRanges.RemoveRanges(availableRanges)
+	available := b.src.Available(shardsTimeRanges)
+
+	remaining := make(map[uint32]xtime.Ranges)
+	for shard, ranges := range shardsTimeRanges {
+		availableRanges, ok := available[shard]
+		if !ok {
+			remaining[shard] = ranges
+			continue
+		}
+
+		remainingRanges := ranges.RemoveRanges(availableRanges)
+		if !remainingRanges.IsEmpty() {
+			remaining[shard] = remainingRanges
+		}
+	}
 
 	var (
-		wg                              sync.WaitGroup
-		curResult, nextResult           bootstrap.ShardResult
-		curUnfulfilled, nextUnfulfilled xtime.Ranges
+		wg                     sync.WaitGroup
+		currResult, nextResult bootstrap.Result
+		currErr, nextErr       error
 	)
+	if len(remaining) > 0 &&
+		b.Can(bootstrap.BootstrapParallel) &&
+		b.next.Can(bootstrap.BootstrapParallel) {
+		// If ranges can be bootstrapped now from the next source then begin attempt now
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			nextResult, nextErr = b.next.Bootstrap(remaining)
+		}()
+	}
 
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		nextResult, nextUnfulfilled = bsb.next.Bootstrap(shard, remainingRanges)
-	}()
+	currResult, currErr = b.src.Read(available)
 
-	curResult, curUnfulfilled = bsb.s.ReadData(shard, availableRanges)
 	wg.Wait()
-
-	mergedResults := bsb.mergeResults(curResult, nextResult)
-
-	// If there are some time ranges the current bootstrapper can't fulfill,
-	// pass it along to the next bootstrapper.
-	if !xtime.IsEmpty(curUnfulfilled) {
-		curResult, curUnfulfilled = bsb.next.Bootstrap(shard, curUnfulfilled)
-		mergedResults = bsb.mergeResults(mergedResults, curResult)
+	if err := xerrors.FirstError(currErr, nextErr); err != nil {
+		return nil, err
 	}
 
-	mergedUnfulfilled := mergeTimeRanges(curUnfulfilled, nextUnfulfilled)
-	return mergedResults, mergedUnfulfilled
-}
+	if currResult == nil {
+		currResult = bootstrap.NewResult()
+	}
 
-func (bsb *baseBootstrapper) mergeResults(results ...bootstrap.ShardResult) bootstrap.ShardResult {
-	final := bootstrap.NewShardResult(bsb.opts)
-	for _, result := range results {
-		final.AddResult(result)
+	var (
+		mergedResult         = currResult
+		currUnfulfilled      = currResult.Unfulfilled()
+		firstNextUnfulfilled bootstrap.ShardTimeRanges
+	)
+	if nextResult != nil {
+		// Union the results
+		mergedResult.ShardResults().AddResults(nextResult.ShardResults())
+		// Save the first next unfulfilled time ranges
+		firstNextUnfulfilled = nextResult.Unfulfilled()
+	} else {
+		// Union just the unfulfilled ranges from current and the remaining ranges
+		currUnfulfilled.AddRanges(remaining)
```

Comment:
This only happens if both `nextErr` and `nextResult` are `nil`, which means `remaining` is empty (?)